feat: 新增上报 client metrics 数据

Ambition9186 opened the pull request:
1. 客户端上报连接、心跳、变更、拉取事件到 feed-server
2. cache-service 定时同步 client metrics 数据到 data-service
3. data-service 定时更新客户端离线状态

alkaidchen commented on bcs-services/bcs-bscp/pkg/thirdparty/protobuf/google/api/timestamp.proto:
这个文件在项目中已存在：
pkg/thirdparty/protobuf/google/protobuf/timestamp.proto

alkaidchen commented on bcs-services/bcs-bscp/pkg/dal/bedis/base.go:
```diff
@@ -410,3 +410,122 @@ func (bs *bedis) Healthz() error {
+	return values, nil
+}
+
+// RPop 移出并获取列表的第一个元素
```
前面 LPush 是插入到头部，那这里就应该是从尾部移出并获取一个元素

alkaidchen commented on bcs-services/bcs-bscp/cmd/cache-service/service/cache/event/client_metric.go:
```diff
@@ -0,0 +1,360 @@
+				logs.Infof("start handle client metric data")
+
+				if !cm.state.IsMaster() {
+					logs.V(2).Infof("this is slave, do not need to purge, skip. rid: %s", kt.Rid)
```
这个日志不对

alkaidchen commented on bcs-services/bcs-bscp/pkg/sf-share/types.go:
```diff
@@ -483,3 +527,363 @@ type KvMetaV1 struct {
+	return jsoni.Marshal(v)
+}
+
+// Decode the VersionChangePayload to bytes.
```
// Decode the VersionChangePayload from bytes.

alkaidchen commented on bcs-services/bcs-bscp/pkg/sf-share/types.go:
```diff
@@ -483,3 +527,363 @@ type KvMetaV1 struct {
+	return jsoni.Marshal(p)
+}
+
+// Decode the PullStatusPayload to bytes.
```
// Decode the PullStatusPayload from bytes.

alkaidchen commented on bcs-services/bcs-bscp/cmd/feed-server/service/rpc_sidecar.go:
```diff
@@ -196,6 +197,99 @@ func (s *Service) Messaging(ctx context.Context, msg *pbfs.MessagingMeta) (*pbfs
 		return nil, status.Errorf(codes.PermissionDenied, "no permission to access bscp server")
 	}
 
+	vc := new(sfs.VersionChangePayload)
```
这几个变量都只是在对应的 case 中才被引用，不需要提前声明，在对应的 case 中声明就行了

alkaidchen commented on bcs-services/bcs-bscp/cmd/feed-server/bll/lcache/cilent_metric.go:
```diff
@@ -0,0 +1,86 @@
+// ClientMetric xxx
+type ClientMetric struct {
+	mc     *metric
+	client gcache.Cache
```
gcache 是一个内存缓存库，这里客户端数据不需要在 feed_server 内存缓存，它不会被客户端读取，只用往 cache-service 写入就行了。

所以这里去掉这个 client

alkaidchen commented on bcs-services/bcs-bscp/pkg/sf-share/types.go:
```diff
@@ -483,3 +532,338 @@ type KvMetaV1 struct {
+}
+
+// VersionChangePayload defines sdk version change to send payload to feed server.
+type VersionChangePayload struct {
```
<img width="451" alt="image" src="https://github.com/TencentBlueKing/bk-bcs/assets/43276537/0ccd97ea-bb36-4e90-a6bb-cbe8cd2662ee">
<img width="551" alt="image" src="https://github.com/TencentBlueKing/bk-bcs/assets/43276537/3bfe6473-27ef-4e36-9512-97a4df146767">
<img width="400" alt="image" src="https://github.com/TencentBlueKing/bk-bcs/assets/43276537/8a517c87-0e97-4476-a64e-c3e9a0e93a65">
<img width="643" alt="image" src="https://github.com/TencentBlueKing/bk-bcs/assets/43276537/e29cd714-80d4-4c25-bca8-6977499e939c">

这几个 Payload 结构都是一样或者相近的，看到这几个 payload 在 bscp-go 上报的时机也都差不多，可不可以合成一个上报呢

alkaidchen commented on bcs-services/bcs-bscp/pkg/dal/dao/client.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+// ListClientByTuple Query the client list according to multiple fields in
+func (dao *clientDao) ListClientByTuple(kit *kit.Kit, data [][]interface{}) ([]*table.Client, error) {
```
这个注释给全一点，不然传入一个二维 interface{} 数组作为入参有点迷惑

alkaidchen commented on bcs-services/bcs-bscp/cmd/data-service/service/crontab/sync_client_online_state.go:
```diff
@@ -0,0 +1,126 @@
+}
+
+// Run the sync client online state task
+func (c *ClientOnlineState) Run() {
```
这个函数也需要支持主从，只有主服务负责定时同步客户端状态

Ambition9186 committed:
1. 新增客户端metrics
2. 新增客户端上报连接、心跳、变更、拉取事件到feed-server
3. 新增cache-service定时同步客户端metrics数据到data-service
4. 新增data-service定时更新客户端离线状态
5. 新增客户端类型字段
